p2panda-net-next/src/topic_streams.rs

```diff
@@ -0,0 +1,158 @@
+use tokio::sync::broadcast::Receiver as BroadcastReceiver;
+use tokio::sync::broadcast::error::{RecvError, TryRecvError};
+use tokio::sync::mpsc::Sender;
+use tokio::sync::mpsc::error::SendError;
+
+use crate::TopicId;
+use crate::actors::subscription::ToSubscription;
+use crate::network::{FromNetwork, ToNetwork};
+use crate::utils::with_suffix;
+
+#[derive(Debug, Error)]
+pub enum TopicStreamError<T> {
+    #[error(transparent)]
+    Send(#[from] SendError<T>),
+
+    #[error(transparent)]
+    Recv(#[from] RecvError),
+
+    #[error("subscription actor is not available to create topic subscription")]
+    SubscriptionNotAvailable,
+}
+
+/// A handle to an ephemeral messaging stream.
+///
+/// The stream can be used to publish messages or to request a subscription.
+pub struct EphemeralTopicStream {
+    topic_id: TopicId,
+    to_topic_tx: Sender<ToNetwork>,
+    public_key_suffix: String,
+}
+
+impl EphemeralTopicStream {
+    /// Returns a handle to an ephemeral messaging stream.
+    pub(crate) fn new(
+        topic_id: TopicId,
+        to_topic_tx: Sender<ToNetwork>,
+        public_key_suffix: String,
+    ) -> Self {
+        Self {
+            topic_id,
+            to_topic_tx,
+            public_key_suffix,
+        }
+    }
+
+    /// Publishes a message to the stream.
+    pub async fn publish(
+        &self,
+        bytes: impl Into<Vec<u8>>,
+    ) -> Result<(), TopicStreamError<Vec<u8>>> {
+        self.to_topic_tx.send(bytes.into()).await?;
+
+        Ok(())
+    }
+
+    /// Subscribes to the stream.
+    ///
+    /// The returned `EphemeralTopicStreamSubscription` provides a means of receiving messages from
+    /// the stream.
+    pub async fn subscribe(
+        &self,
+    ) -> Result<EphemeralTopicStreamSubscription, TopicStreamError<()>> {
+        // Get a reference to the subscription actor.
+        if let Some(subscription_actor) =
+            registry::where_is(with_suffix("subscription", &self.public_key_suffix))
+        {
+            let actor: ActorRef<ToSubscription> = subscription_actor.into();
+
+            // Ask the subscription actor for an ephemeral stream subscriber.
+            let subscription = call!(
+                actor,
+                ToSubscription::ReturnEphemeralSubscription,
+                self.topic_id
+            )
+            .map_err(|_| TopicStreamError::SubscriptionNotAvailable)?;
+
+            Ok(subscription)
+        } else {
+            Err(TopicStreamError::SubscriptionNotAvailable)
+        }
+    }
+
+    /// Returns the topic ID of the stream.
+    pub fn topic_id(&self) -> TopicId {
+        self.topic_id
+    }
+
+    /// Unsubscribes from the ephemeral messaging stream.
+    pub fn unsubscribe(self) -> Result<(), TopicStreamError<()>> {
+        if let Some(subscription_actor) =
+            registry::where_is(with_suffix("subscription", &self.public_key_suffix))
+        {
+            let actor: ActorRef<ToSubscription> = subscription_actor.into();
```

Comment:
This is slightly annoying boilerplate which is repeated here and there, it would make sense to factor it into a dedicated method.

Reply:
nit: It could make sense to use more "guard clauses" or "early returns" to avoid nesting (can be applied to other parts of this PR as well)